### PR TITLE
Update GitHub Actions to leverage pre-commit hook framework and integrate flake8 hook

### PR DESCRIPTION
**Pull request by jcfr**
This pull-request adds workflow leveraging pre-commit, a framework for
managing and maintaining pre-commit hooks.

See https://pre-commit.com/ and https://github.com/pre-commit/action

Running the pre-commit locally can be done using the following
command:

```
pre-commit run --all-files
```

The hooks enabled in the "pre-commit-config.yaml" configuration
file are:

* `check-added-large-files`
  Prevent giant files from being committed.

  Set the size to 1024 to match the value hard-coded in
  https://github.com/Slicer/Slicer/blob/hooks/pre-commit#L125

* `check-case-conflict`
  Check for files with names that would conflict on a case-insensitive
  filesystem like MacOS HFS+ or Windows FAT.

* `check-merge-conflict`
  Check for files that contain merge conflict strings.

* `check-symlinks`
  Checks for symlinks which do not point to anything.

----

To allow a gradual improvement of the code base while carefully reviewing
each changes, exceptions for all issues reported by locally running the
following command are listed in `.flake8` configuration files:

```
  pre-commit run --all-files
```

References:
* https://www.flake8rules.com/
* https://github.com/PyCQA/flake8-bugbear#list-of-warnings

----

It was adapted from prior work done by @henryiii
in https://github.com/scikit-build/scikit-build

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+jobs:
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: actions/setup-python@v2
+
```
[the remark this answers is not in the document]
Comment (henryiii) [reply .github/workflows/lint.yml:19]: These are both v3 now, by the way. For 99% of linting and formatters, I've found "any newer Python" is fine. PyLint is the only one I think might matter occasionally. But it is also fine to match.

**Comment by henryiii** [comment]
FYI, another good reference: https://scikit-hep.org/developer/style

**Comment by jcfr** [comment]
@jamesobutler @lassoan @pieper @sjh26 Unless you have any additional comments, I plan on integrating this later this afternoon.

There are additional cleanup coming up as well :rocket:

**Comment by jcfr** [comment]
Thanks everyone for the review :pray:

**Comment by lassoan** [comment]
Thanks @jcfr for working on this. Overall, how the development workflow is going to change?

Is linting going to be performed for all PR?

Is it going to be performed locally automatically by the pre-commit hook? How that can be set up on Windows? I don't see any developer documentation file changes.

**Comment by jcfr** [comment]
For now, this will only report info to developers after the PR is submitted.

There are no changes in the hook setup locally yet.

I will submit a follow up PR explaining how to manually run the pre-commit checks locally.

**Pull request merged**